### ISSUE TITLE
Possible bug in `Tensor::matmul`

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#811]
I was implementing LoRA for Embedding, when I realized that the `.matmul` method was giving an error where Pytorch's does not. Is this a bug?

### Rust (Candle)
```rust
let a = /*get some tensor with dims (2,4,1)*/
let b = /*get some tensor with dims (1,3)*/
let res = a.matmul(b)?;
```
Error: ```thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ShapeMismatchBinaryOp { lhs: [2, 4, 1], rhs: [1, 3], op: "matmul" }'```
### Python (Pytorch)
```rust
import torch
a=torch.randn((2,4,1))
b=torch.randn((1,3))
res = a @ b
```
No error. Shape of `res` is (2,4,3).

**Comment by LaurentMazare** [comment]
Binary ops in candle are not broadcasting by default as it's very error-prone. You can either call `broadcast_matmul` or broadcast `b` directly to the proper batch dim (probably the safest).

**Comment by LaurentMazare** [comment]
Yes `u32` is sadly not supported in this backend at the moment, you probably want `F32`, `F16`, or `F64`, the error is really poor I'll improve this.

**Comment by EricLBuehler** [comment]
Ok, that is too bad. I need an integral type for Embedding to work.

**Comment by LaurentMazare** [comment]
Why would you need a matmul rather than something like `index-select`? (see how [candle-nn::embeddings](https://github.com/huggingface/candle/blob/main/candle-nn/src/embedding.rs) work)

**Comment by EricLBuehler** [comment]
I followed what https://github.com/microsoft/LoRA/blob/main/loralib/layers.py does. How would I use that instead?

**Comment by LaurentMazare** [comment]
These seem to be `f32` tensors in this code or am I missing something? (I wouldn't see a point about using a product between two `u32` matrixes as you wouldn't be able to backprop through them)

**Comment by EricLBuehler** [comment]
Yes, I realized that and just fixed it :)

**Comment by LaurentMazare** [comment]
Ah great, all good then!

**Comment by EricLBuehler** [comment]
Thank you!

**Comment by LaurentMazare** [comment]
#813 should make the original error clearer, thanks for reporting it.

**Comment by EricLBuehler** [comment]
Great, I just checked it out.
